packages/@aws-cdk/aws-applicationautoscaling/lib/scalable-target.ts

```diff
@@ -0,0 +1,211 @@
+   */
+  public readonly scalableTargetId: string;
+
+  /**
+   * The role used to give AutoScaling permissions to your resource
+   */
+  public readonly role: iam.IRole;
+
+  private readonly actions = new Array<cloudformation.ScalableTargetResource.ScheduledActionProperty>();
+
+  constructor(parent: cdk.Construct, id: string, props: ScalableTargetProps) {
+    super(parent, id);
+
+    if (props.maxCapacity < 0) {
+      throw new RangeError(`maxCapacity cannot be negative, got: ${props.maxCapacity}`);
+    }
+    if (props.minCapacity < 0) {
+      throw new RangeError(`minCapacity cannot be negative, got: ${props.minCapacity}`);
+    }
+    if (props.maxCapacity < props.minCapacity) {
+      throw new RangeError(`minCapacity (${props.minCapacity}) should be lower than maxCapacity (${props.maxCapacity})`);
+    }
+
+    this.role = props.role || new iam.Role(this, 'Role', {
+      assumedBy: new iam.ServicePrincipal('application-autoscaling.amazonaws.com')
+    });
+
+    const resource = new cloudformation.ScalableTargetResource(this, 'Resource', {
+      maxCapacity: new cdk.Token(() => props.maxCapacity),
+      minCapacity: new cdk.Token(() => props.minCapacity),
+      resourceId: props.resourceId,
+      roleArn: this.role.roleArn,
+      scalableDimension: props.scalableDimension,
+      scheduledActions: this.actions,
+      serviceNamespace: props.serviceNamespace
+    });
+
+    this.scalableTargetId = resource.scalableTargetId;
+  }
+
+  /**
+   * Add a policy statement to the role's policy
+   */
+  public addToRolePolicy(statement: iam.PolicyStatement) {
+    this.role.addToPolicy(statement);
+  }
+
+  /**
+   * Scale out or in based on time
+   */
+  public scaleOnSchedule(id: string, action: ScalingSchedule) {
+    if (action.minCapacity === undefined && action.maxCapacity === undefined) {
+      throw new Error(`You must supply at least one of minCapacity or maxCapacity, got ${JSON.stringify(action)}`);
+    }
+    this.actions.push({
+      scheduledActionName: id,
+      schedule: action.schedule,
+      startTime: action.startTime,
+      endTime: action.endTime,
+      scalableTargetAction: {
+        maxCapacity: action.maxCapacity,
+        minCapacity: action.minCapacity
+      },
+    });
+  }
+
+  /**
+   * Scale out or in, in response to a metric
+   */
+  public scaleOnMetric(id: string, props: BasicStepScalingPolicyProps) {
+    return new StepScalingPolicy(this, id, { ...props, scalingTarget: this });
+  }
+
+  /**
+   * Scale out or in in order to keep a metric around a target value
+   */
+  public scaleToTrackMetric(id: string, props: BasicTargetTrackingScalingPolicyProps) {
+    return new TargetTrackingScalingPolicy(this, id, { ...props, scalingTarget: this });
+  }
+}
+
+/**
+ * A scheduled scaling action
+ */
+export interface ScalingSchedule {
+  /**
+   * When to perform this action.
+   *
+   * Support formats:
```

Comment:
Can we not type those? Feels like the expression domain is pretty tight and we could use real language types instead of strings...

Reply:
Yyyyeaaahhh... I was going back and forth on that. This has the advantage of being simple (as in: being close to what people expect and works the same as the other APIs). My plan to paper over that was to make helper classes that will generate this string for you. Seems better than to force people to instantiate a class if they already know what string they want?